### PR TITLE
add HSP IAM audit capabilities

### PR DESCRIPTION
**Pull request by EriksonBahr**
## Description

Add HSDP IAM audit capabilities to oauth2 proxy

## Motivation and Context

HSDP IAM does not provide important audit capabilities out of the box such as successful login and some key auth errors.

## How Has This Been Tested?

Tested locally with:

```sh
export OAUTH2_PROXY_ENABLE_AUDIT='true'
export OAUTH2_PROXY_AUDIT_URL='url'
export OAUTH2_PROXY_AUDIT_PRODUCT_NAME='Reporting'
export OAUTH2_PROXY_AUDIT_PRODUCT_KEY='product key'
export OAUTH2_PROXY_AUDIT_SHARED_KEY='shared key'
export OAUTH2_PROXY_AUDIT_SECRET_KEY='secret'
```

Example of the payload:

```
{
  "resourceType": "AuditEvent",
  "event": {
    "type": {
      "system": "http://hl7.org/fhir/ValueSet/audit-event-type",
      "version": "1",
      "code": "110114",
      "display": "User Authentication"
    },
    "action": "E",
    "dateTime": "2024-07-03T16:34:44Z",
    "outcome": "0",
    "outcomeDesc": "Success"
  },
  "participant": [
    {
      "userId": { "value": "admin_t952@philips.com" },
      "altId": "admin_t952",
      "requestor": true
    }
  ],
  "source": {
    "identifier": {
      "type": {
        "system": "http://hl7.org/fhir/ValueSet/audit-source-type",
        "code": "4",
        "display": "Application Server"
      },
      "value": "admin_t952@philips.com"
    },
    "type": [
      {
        "system": "http://hl7.org/fhir/security-source-type",
        "code": "1",
        "display": "End-user display device, diagnostic device."
      }
    ],
    "extension": [
      {
        "url": "/pics/reporting?procedure.identifier=https%3A%2F%2Fwww.philips.com%2Fidentifiers%2FProcedureIdentifier|ACC_20240625221043040",
        "extension": [
          { "url": "applicationName", "valueString": "Reporting" },
          { "url": "applicationVersion", "valueString": "1" },
          { "url": "serverName", "valueString": "oauth2proxy" },
          { "url": "componentName", "valueString": "oauth2proxy" },
          {
            "url": "productKey",
            "valueString": "<omited>"
          },
          {
            "url": "tenant",
            "valueString": "3b64ff86-8ecf-47cb-8431-bedf41318e43"
          }
        ]
      }
    ]
  }
}

```

## Checklist:

- [ ] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [ ] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: pkg/middleware/cookie_refresh.go [hunk 1]
```diff
@@ -41,6 +41,7 @@ func (cr *cookieRefresh) refreshCookie(next http.Handler) http.Handler {
 		cookie, err := req.Cookie(cr.CookieRefreshName)
 		if err != nil {
 			logger.Errorf("SSO Cookie Refresher - Could find '%s' cookie in the request: %v", cr.CookieRefreshName, err)
+			next.ServeHTTP(rw, req)
```
Comment (evozniak) [inline pkg/middleware/cookie_refresh.go:44]: If the refresh fails should we continue processing middlewares?
Is there something wrong with the token refresher?

File: pkg/audit/audit_client.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+			DateTime:    time.Now().UTC().Format(time.RFC3339),
+			Outcome:     outcomeCode,
+			OutcomeDesc: outcomeDesc},
+
+		Participant: []*Participant{
+			{AltID: ss.User, UserID: UserID{Value: ss.Email}, Name: ss.PreferredUsername, Requestor: true}},
+		Source: Source{
+			Identifier: Identifier{
+				Type: &Coding{
+					System:  "http://hl7.org/fhir/ValueSet/identifier-type",
```
Comment (evozniak) [inline pkg/audit/audit_client.go]: Nit: I couldn't find code 4 on this value set, am I missing something?

File: pkg/audit/audit_client.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	}
+	if resp.StatusCode() != 201 {
+		log.Println("Not able to send the audit message ", resp)
+		return fmt.Errorf("not able to persist audit, audit server returned %v", resp.StatusCode())
+	}
+	return nil
+}
+
+func (c *ClientOpts) Validate() error {
+	if strings.TrimSpace(c.ProductName) == "" || strings.TrimSpace(c.ProductKey) == "" || strings.TrimSpace(c.SecretKey) == "" || strings.TrimSpace(c.SharedKey) == "" {
```
Comment (evozniak) [inline pkg/audit/audit_client.go]: Isn't it better to validate each value individually?

File: pkg/audit/audit_client.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+		return errors.New("the audit is enabled and therefore the audit product name, audit key, audit secret key or audit shared key are required (however found empty)")
```
Comment (evozniak) [inline pkg/audit/audit_client.go]: You created audit_error.go, why this error is not there?

**Review by evozniak — DISMISSED** [review]
Minor comments, looking good.

File: pkg/audit/audit_client.go (hunk 4 above)
Comment (EriksonBahr) [reply pkg/audit/audit_client.go]: this is more a setup error and the others are actually runtime errors

File: pkg/middleware/cookie_refresh.go (hunk 1 above)
Comment (EriksonBahr) [reply pkg/middleware/cookie_refresh.go:44]: we should keep going, this is the last middleware so this change doesn't actually change the behavior

File: pkg/audit/audit_client.go (hunk 2 above)
Comment (EriksonBahr) [reply pkg/audit/audit_client.go]: I think I copied the wrong one, let me see

File: pkg/audit/audit_client.go (hunk 3 above)
Comment (EriksonBahr) [reply pkg/audit/audit_client.go]: it is

File: pkg/audit/audit_client.go (hunk 2 above)
Comment (EriksonBahr) [reply pkg/audit/audit_client.go]: fixed

File: pkg/audit/audit_client.go (hunk 3 above)
Comment (EriksonBahr) [reply pkg/audit/audit_client.go]: fixed

**Review by evozniak — APPROVED** [review]

**Pull request merged**